### PR TITLE
[ASCII-858] Separate pulumi logic from VM client

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?
This PR:
-  makes a clear distinction between the code that interacts with a remote VM and the code that initializes a vm client from a pulumi stack. 
- makes a clear distinction between the code that interacts with a remote Agent and the code that initializes an agent client from a pulumi stack.
- Add more explicit comments
- Remove the restriction that a field in a Env must implement pulumiStackInitializer

It is recommended to review this PR commit per commit.

Note: A followup PR is require to remove os.OS instance from `VMClient` and `AgentClient`.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by pducolin — APPROVED** [review]
Minor comment on keeping the VM interface implementation to `VMClient` rather than to both `VMClient` and `PulumiVM`, that implicitly implements it thanks to embedding a `VMCient`

File: test/new-e2e/pkg/utils/e2e/client/pulumi_stack_vm.go [hunk 1]
```diff
@@ -16,8 +16,9 @@ import (
 )
 
 var _ stackInitializer = (*PulumiStackVM)(nil)
+var _ VM = (*PulumiStackVM)(nil)
 
-// PulumiStackVM is a type to help creating a VM client connecting to a
+// PulumiStackVM is a type to help creating a VM (vm.go) connecting to a
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/pkg/utils/e2e/client/pulumi_stack_vm.go]: This comment was updated in a later commit.

File: test/new-e2e/pkg/utils/e2e/client/pulumi_stack_vm.go [hunk 2]
```diff
@@ -16,8 +16,9 @@ import (
 )
 
 var _ stackInitializer = (*PulumiStackVM)(nil)
+var _ VM = (*PulumiStackVM)(nil)
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/pkg/utils/e2e/client/pulumi_stack_vm.go:19]: Unfortunately, embed struct is not inheritance. See [this code](https://go.dev/play/p/f_J9424AtDy) as an example why it doesn't work.

**Review by KevinFairise2 — APPROVED** [review]

File: test/new-e2e/pkg/utils/e2e/client/pulumi_stack_initializer_test.go [hunk 3]
```diff
@@ -33,5 +33,5 @@ type DoesNotImplementInterfaceEnv struct {
 }
 
 func TestDoesNotImplementInterfaceEnv(t *testing.T) {
-	require.Error(t, CheckEnvStructValid[DoesNotImplementInterfaceEnv]())
+	require.NoError(t, CheckEnvStructValid[DoesNotImplementInterfaceEnv]())
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/pkg/utils/e2e/client/pulumi_stack_initializer_test.go:36]: This PR change this behavior. See PR description `Remove the restriction that a field in a Env must implement pulumiStackInitializer`

**Review by pducolin — APPROVED** [review]

**Pull request merged**